web/pandas/pdeps/0005-no-default-index-mode.md

```diff
@@ -0,0 +1,189 @@
+# PDEP-5: No-default-index mode
+
+- Created: 14 November 2022
+- Status: Draft
+- Discussion: [#49693](https://github.com/pandas-dev/pandas/pull/49693)
+- Author: [Marco Gorelli](https://github.com/MarcoGorelli)
+- Revision: 1
+
+## Abstract
+
+The suggestion is to add a `mode.no_default_index` option which, if enabled,
+would ensure:
+- if a ``DataFrame`` / ``Series`` is created, then by default it won't have an ``Index``;
+- nobody will get an ``Index`` unless they ask for one - this would affect the default behaviour of ``groupby``, ``value_counts``, ``pivot_table``, and more.
+
+This option would not be the default. Users would need to explicitly opt-in to it, via ``pd.set_option('mode.no_default_index', True)``, via ``pd.option_context``, or via the ``PANDAS_NO_DEFAULT_INDEX`` environment variable.
+
+## Motivation and Scope
+
+The Index can be a source of confusion and frustration for pandas users. For example, let's consider the inputs
+
+```python
+In [37]: ser1 = df.groupby('sender')['amount'].sum()
+
+In [38]: ser2 = df.groupby('receiver')['amount'].sum()
+
+In [39]: ser1
+Out[39]:
+sender
+1    10
+2    15
+3    20
+5    25
+Name: amount, dtype: int64
+
+In [40]: ser2
+Out[40]:
+receiver
+1    10
+2    15
+3    20
+4    25
+Name: amount, dtype: int64
+```
+. Then:
+
+- it can be unexpected that summing `Series` with the same length (but different indices) produces `NaN`s in the result (https://stackoverflow.com/q/66094702/4451315):
+
+  ```python
+  In [41]: ser1 + ser2
+  Out[41]:
+  1    20.0
+  2    30.0
+  3    40.0
+  4     NaN
+  5     NaN
+  Name: amount, dtype: float64
+  ```
+
+- concatenation, even with `ignore_index=True`, still aligns on the index (https://github.com/pandas-dev/pandas/issues/25349):
+
+    ```python
+    In [42]: pd.concat([ser1, ser2], axis=1, ignore_index=True)
+    Out[42]:
+          0     1
+    1  10.0  10.0
+    2  15.0  15.0
+    3  20.0  20.0
+    5  25.0   NaN
+    4   NaN  25.0
+    ```
+
+- it can be frustrating to have to repeatedly call `.reset_index()` (https://twitter.com/chowthedog/status/1559946277315641345):
+
+    ```python
+    In [45]: df.value_counts(['sender', 'receiver']).reset_index().rename(columns={0: 'count'})
+    Out[45]:
+       sender  receiver  count
+    0       1         1      1
+    1       2         2      1
+    2       3         3      1
+    3       5         4      1
+    ```
+
+With this option enabled, users who don't want to worry about indices wouldn't need to.
+
+## Detailed Description
+
+This would require 3 steps:
+1. creation of a ``NoIndex`` object, which would be a subclass of ``RangeIndex`` on which
```

Comment:
Curious why `NoIndex` be a subclass of `RangeIndex`? Does it just make things easier to implement?

IMO it _feels_ like `NoIndex` should be an abstract-like class that Index and RangeIndex should subclass (and implement the "index-like" functionality

Reply:
Or why not just `None`? (i.e. `df.index is None`)

In any case I agree with @mroeschke that it seems strange to make this a subclass of RangeIndex, since we explicitly _don't_ want to have it act as a RangeIndex

Reply:
I expect that under the hood (e.g. at the Manager level) we'd still have an Index object (until/unless we refactor to get the indexes out of the Managers) and a NoIndex object makes sense there.

Reply:
Yes, but that could be an internal implementation detail (to avoid having to check internally for `if index is not None: ` too often). I think what matters here is what we want for the user. In the public `.index` property, we could convert a "NoIndex" object into `None`, if we think that is the better user API.

(although in the internals at least we don't do much with the index, we mostly pass it through, and passing through None could also work; I suppose it is mostly outside of the internals that a NoIndex object could make an initial implementation easier (less places to update))

Reply:
> Not sure I follow the suggestion about it being an abstract class:

Sorry I think "abstract-like" was probably a loaded term to use: I meant more akin to a skeleton class (that you alluded to) where you can define the necessary methods and raise a `NotImeplementedError` to "disable" indexing-like behavior

Reply:
> So rather than re-engineering `_set_with_engine` to not use `self.index` if `index` is `None`, I thought it'd be simpler to have some `NoIndex` object which overrides `get_loc`
> 
> Then, with `get_loc`, `append`, and `join` overridden, very little would need changing outside of the `NoIndex` class itself

For me, I would say this is still an implementation detail. We could decide to internally use for example `._index` instead of `.index`, which would be ensured to return an Index (sub)class instance, so that this can be implemented this way without too much changes (apart from adding the underscore). And this way the public property could still return `None`. 

For the public API, the question is: would it be useful for the user to have a NoIndex object? (to be clear, I am not necessarily saying the answer is definitly "no", we should probably think through some use cases) How would the user make use of this? 
And do we want that the user starts relying on the fact that they can still call all the Index methods on this NoIndex object?

(and for this PDEP, I don't think we necessarily have to agree on the exact implementation approach, but I would like to see a concrete API proposal for the public API)